From 6d087d5dd757fcfac41a4b77f2304c9aa015888c Mon Sep 17 00:00:00 2001
From: Emmanuel Kasper <manu@debian.org>
Date: Sat, 11 Jan 2020 11:34:59 +0100
Subject: [PATCH] Allow files without the executable bit to be run by mksh

Please note that this breaks posix, as running a binary without the executable bit should fail and return the 126 code when running via sh -x.
---
 mksh/edit.c  | 4 ++--
 mksh/exec.c  | 6 +++---
 mksh/funcs.c | 2 +-
 3 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/edit.c b/edit.c
index c231af1..90935c9 100644
--- a/edit.c
+++ b/edit.c
@@ -827,10 +827,10 @@ glob_path(int flags, const char *pat, XPtrV *wp, const char *lpath)
 		glob_str(Xstring(xs, xp), wp, true);
 		newsize = XPsize(*wp);
 
-		/* Check that each match is executable... */
+		/* Check that each match is at least readble ... */
 		words = (char **)XPptrv(*wp);
 		for (i = j = oldsize; i < newsize; i++) {
-			if (ksh_access(words[i], X_OK) == 0) {
+			if (ksh_access(words[i], R_OK) == 0) {
 				words[j] = words[i];
 				if (!(flags & XCF_FULLPATH))
 					memmove(words[j], words[j] + pathlen,
diff --git a/exec.c b/exec.c
index 8231b54..f0b433a 100644
--- a/exec.c
+++ b/exec.c
@@ -887,7 +887,7 @@ scriptexec(struct op *tp, const char **ap)
 
 	sh = str_val(global(TEXECSHELL));
 	if (sh && *sh)
-		sh = search_path(sh, path, X_OK, NULL);
+		sh = search_path(sh, path, R_OK, NULL);
 	if (!sh || !*sh)
 		sh = MKSH_DEFAULT_EXECSHELL;
 
@@ -1209,7 +1209,7 @@ findcom(const char *name, int flags)
 	if (!tp && (flags & FC_PATH) && !(flags & FC_DEFPATH)) {
 		tp = ktsearch(&taliases, name, h);
 		if (tp && (tp->flag & ISSET) &&
-		    ksh_access(tp->val.s, X_OK) != 0) {
+		    ksh_access(tp->val.s, R_OK) != 0) {
 			if (tp->flag & ALLOC) {
 				tp->flag &= ~ALLOC;
 				afree(tp->val.s, APERM);
@@ -1234,7 +1234,7 @@ findcom(const char *name, int flags)
 		}
 		npath.ro = search_path(name,
 		    (flags & FC_DEFPATH) ? def_path : path,
-		    X_OK, &tp->u2.errnov);
+		    R_OK, &tp->u2.errnov);
 		if (npath.ro) {
 			strdupx(tp->val.s, npath.ro, APERM);
 			if (npath.ro != name)
diff --git a/funcs.c b/funcs.c
index a42c310..fac4f34 100644
--- a/funcs.c
+++ b/funcs.c
@@ -889,7 +889,7 @@ c_alias(const char **wp)
 			}
 			/* ignore values for -t (AT&T ksh does this) */
 			newval = tflag ?
-			    search_path(alias, path, X_OK, NULL) :
+			    search_path(alias, path, R_OK, NULL) :
 			    val;
 			if (newval) {
 				strdupx(ap->val.s, newval, APERM);
-- 
2.24.0

